feat: `core` refactor

dalechyn opened the pull request:
Problem: The current state of the core is not reliable for use. Some of the parameters of the functions are simply ignored (faced that with `apiRequest` previously), some are counter-intuitive and use repetitive naming. Matter fact, ~30% code of the classes is its initialization, and sometimes the actual business logic are just two or even a single function.

Solution:
- Move away from writing custom classes on top of the other wherever possible and rely on the `@abstract-money/cli` generated modules as much as possible.
- Provide intuitive and declarative naming to the methods to be used.
- De-classify classes that should not have been classes and reduce complexity.
- To eliminate confusion between `AnsAsset` and `Asset` (which has an `amount` field), introduce `Token` type.

  The list will be populated as the refactor goes.

dalechyn committed:
feat: `core` refactor

Problem: The current state of the core is not reliable for use.
Some of the parameters of the functions are simply ignored (faced that
with `apiRequest` previously), some are counter-intuitive and use
repetitive naming. Matter fact, ~30% code of the classes is its
initialization, and sometimes the actual business logic are just two or
even a single function.

Solution:
- Move away from writing custom classes on top of the other wherever
  possible and rely on the `@abstract-money/cli` generated modules as
  much as possible.
- Provide intuitive and declarative naming to the methods to be used.
- De-classify classes that should not have been classes and reduce
  complexity.
- To eliminate confusion between `AnsAsset` and `Asset` (which has an
  `amount` field), introduce `Token` type.

  The list will be populated as the refactor goes.

dalechyn committed:
feat: clients, separation of concerns

I figured out that we actually have 5 clients:
- base client – knows only the `apiUrl`
- public client - knows the `apiUrl` and `CosmWasmClient`
- wallet client - knows the `apiUrl` and `SigningCosmWasmClient`
- account-public client - knows it's accountId, `apiUrl` and `CosmWasmClient`
- account-wallet client - knows it's accountId, `apiUrl` and `SigningCosmWasmClient`

The commit is still WIP, yet mostly it's just the functionality of the abstract
account that is left to be implemented.

Following folders were created to manage the codebase effectively:
- `actions` – contains all actions that clients are able to perform
|  - `public` – contains all public client actions
|  - `wallet` – contains all wallet client actions
|  - `account-public` – contains all account-public client actions
|  - `account-wallet` – contains all account-wallet client actions
|  ... - base client actions
|  Some of them are grouped by module groups, such as `ans`, `account-factory`, `manager` etc.
- `utils` – contains all utility data-transformation, formatting functions
– `clients` – contains the clients implementations
|  - `decorators` – contains the decorators to inject the client data and bind those to clients
- `types` – contains types

With such a structure, we're clearly separating the business logic from the utilitary logic that
mostly concerns different data transormations.

The focus here is also to keep the functions with the most minimal API as possible, so that hardly
re-uses `ts-codegen` generated types whenever possible.

The API might not look user-friendly at the moment, but later on we will provide aliases for more
user-friendly function names.

adairrr commented on packages/core/src/actions/account-factory/get-account-factory-address-from-api.ts:
```diff
@@ -0,0 +1,27 @@
+) {
+  const deploymentData = await request(
+    apiUrl,
+    gql(/* GraphQL */ `
+      query AccountFactoryDeployment($chain: ID!) {
+        version
+        deployment(chain: $chain) {
+          accountFactory
+        }
+        chainInfo(chain: $chain) {
+          rpcUrl
+        }
+      }
+    `),
+    {
+      chain: chainId,
+    },
+  )
```
We should not request extraneous information besides the `deployment.accountFactory`.

adairrr replied:
Same for the other queries .

dalechyn replied:
agree, that's unnecessary.

adairrr commented on packages/core/src/actions/get-accounts-of-owner-from-api.ts:
```diff
@@ -0,0 +1,43 @@
+        chainName
+        sequence
+        trace
+      }
```
We only need this information from the API in this example.

adairrr commented on packages/core/src/utils/tokens/get-tokens-from-ans.ts:
```diff
@@ -0,0 +1,51 @@
+  const result = await request(
+    apiURL,
+    gql(
+      `
```
```suggestion
    gql(/* GraphQL */
      `
```

adairrr commented on packages/core/src/utils/tokens/get-tokens-from-ans.ts:
```diff
@@ -0,0 +1,51 @@
+import { CW20Token } from './cw20/cw20-token'
+import { NativeToken } from './native/native-token'
+
+export async function getTokensFromAns(
```
Maybe `getTokensByNamesFromAns` would be more accurate

dalechyn replied:
> Maybe `getTokensByNamesFromAns` would be more accurate

Indeed but I would suggest specifying the criteria only when there are >1 methods that pull data from a single source.

Ex: `getTokensByNamesFromAns`, `getTokensBySomethingElseFromAns` that would totally make sense.

adairrr replied:
I mean as opposed to just listing the tokens, for which this method name makes sense.

dalechyn replied:
Changed to `get-ans-tokens-from-api.ts`

adairrr commented on packages/core/src/utils/tokens/get-token-from-ans.ts:
```diff
@@ -0,0 +1,54 @@
+  ansName: string,
+  chain: string,
+  apiURL: string,
+) {
```
Would rather this method call into the `getTokensFromAns` method below and take the first, as duplicating the functionality is going to be less productive if we have to make changes later.

dalechyn replied:
will check it out, sure.

adairrr commented on packages/core/src/utils/chain-registry/chains.ts:
```diff
@@ -0,0 +1,152 @@
+  // remove harpoon 4 in favor of manual entry
+  ...chainRegistryChains.filter(
+    ({ chain_id }) =>
+      !['uni-6', 'pisco-1', 'archway-1', 'neutron-1'].includes(chain_id),
```
```suggestion
      ![NIBIRU_TESTNET_CHAIN.chain_id].includes(chain_id),
```

we should only remove the ones we've overridden.

dalechyn replied:
we're actually not overriding it since there is no such chain in the `chain-registry`, but extending.

anyway this is incorrect.

adairrr replied:
Even IF the chain were to be included, with this logic it wouldn't be included twice.

dalechyn replied:
such would not be normally possible unless we don't double check the `chain-registry` by accident when upgrading one.

the current version of `chain-registry` does not have nibiru testnet.

I will put the check anyway.

adairrr commented on packages/core/src/utils/chain-registry/chain-id-to-name.ts:
```diff
@@ -0,0 +1,12 @@
+import { chains } from './chains'
+
+export function chainIdToName(chainId: string) {
+  if (chainId === 'pisco-1') return 'terra2testnet'
```
We might consider pulling out `CHAINID_TO_NAME_OVERRIDES` for clarity.

dalechyn replied:
will make sense when we will have multiple chains to override.

adairrr commented on packages/core/src/utils/chain-registry/assets.ts:
```diff
@@ -0,0 +1,27 @@
+}
+
+export const assets: typeof cosmosAssets = [
+  ...cosmosAssets,
```
We should filter out the assets that we've overridden.

adairrr commented on packages/core/src/utils/assets/encode-assets-transfer-msgs.ts:
```diff
@@ -0,0 +1,48 @@
+import { NativeAsset, isNativeAsset } from './native/native-asset'
+import { nativeAssetToCoin } from './native/native-asset-to-coin'
+
+export function encodeAssetsTransfersMsgs<
```
I personally liked the `Asset` class because it mirrored the experience in the smart-contracts working with the same struct impls (vs classes).

The functional code should be a lot more testable, however.

adairrr commented on packages/core/src/types/utils.ts:
```diff
@@ -0,0 +1,35 @@
+/**
+ * @description Combines members of an intersection into a readable type.
+ *
+ * @see {@link https://twitter.com/mattpocockuk/status/1622730173446557697?s=20&t=NdpAcmEFXY01xkqU3KO0Mg}
```
Love Matt Pocock

adairrr commented on packages/core/src/utils/account-id/account-id-to-parameter.ts:
```diff
@@ -0,0 +1,6 @@
+import { VersionControlTypes } from 'src/codegen/abstract'
+
+// TODO: Figure out where is this used
```
This would be used internally when calling into version control.

adairrr commented on packages/core/src/utils/account-id/account-id.ts:
```diff
@@ -0,0 +1,8 @@
+
+export type AccountId<TChainName extends string = string> = {
+  chainName: TChainName
+} & VersionControlTypes.AccountId
```
This will make `AccountId` have `seq` instead of `sequence` as well... 

I would consider not tying the `AccountId` details directly to the smart-contract API, which is not at all likely to change, but isn't the most dev-friendly IMO.

And since the AccountId was used so often, it was really convenient to have all the helpers defined on the `AccountId` instance as it was passed around the frontend. Importing functions it's not clear what the available actions will have and the code is "fragmented" away from the central implementation (though yes they're in the same directory).

dalechyn replied:
Maybe we should rename the `seq` property in the smart contracts instead?

adairrr commented on packages/core/src/clients/create-public-client.ts:
```diff
@@ -0,0 +1,22 @@
+
+export type PublicClient = Prettify<Client<PublicActions>>
+
+export function createPublicClient(
```
This decorator pattern is fascinating and I haven't seen it in action before. I see it mirrors viem 👀

dalechyn replied:
yep yep, heavily inspired.

adairrr commented on packages/core/src/clients/decorators/public.ts:
```diff
@@ -0,0 +1,62 @@
+import { CosmWasmClient } from '@cosmjs/cosmwasm-stargate'
+import { getAccountFactoryAddress } from '../../actions/public/account-factory/get-account-factory-address.js'
+import { getAccountFactoryQueryClient } from '../../actions/public/account-factory/get-account-factory-query-client'
+import { getAccountFactoryQueryClientFromApi } from '../../actions/public/account-factory/get-account-factory-query-client-from-api'
+import { getAnsHostAddress } from '../../actions/public/ans/get-ans-host-address.js'
+import { getAnsHostQueryClientFromApi } from '../../actions/public/ans/get-ans-host-query-client-from-api.js'
+import { getAnsHostQueryClient } from '../../actions/public/ans/get-ans-host-query-client.js'
+import { getVersionControlModuleData } from '../../actions/public/version-control/get-version-control-module-data.js'
+import { getVersionControlQueryClientFromApi } from '../../actions/public/version-control/get-version-control-query-client-from-api.js'
+import { SliceFirst, SliceFirstTwo } from '../../types/utils.js'
+import { Client } from '../create-client.js'
```
Any reason these are js imports?

dalechyn replied:
nope, will change

adairrr commented on packages/core/src/api/apiRequest.ts:
Curious what you mean by 
```
Some of the parameters of the functions are simply ignored (faced that
with `apiRequest` previously)
```
... Have never experienced this personally, though the method was redundant in playing around with it now.

dalechyn replied:
![Screenshot 2024-01-16 at 14 19 28](https://github.com/AbstractSDK/abstract.js/assets/35642018/4c98764b-81b5-49dd-be01-11cd1776f997)
https://github.com/AbstractSDK/abstract.js/commit/1a29629da2917e1f614e3fe32d0dc74625562da0

adairrr replied:
Oof yeah that case was certainly never tested. 

Thought you meant of the `apiRequest` method specifically. thanks for clarifying.

adairrr commented on packages/core/src/actions/public/version-control/get-version-control-module-data.ts:
```diff
@@ -0,0 +1,58 @@
+      )
+      const firstInstantiation = instantiations[0]
+      if (!firstInstantiation) {
+        console.log(
```
```suggestion
        console.warn(
```

adairrr commented on packages/core/src/actions/public/parse-create-account-execute-result.ts:
```diff
@@ -0,0 +1,26 @@
+import { chainIdToName } from '../../utils/chain-registry'
+import { findAbstractAttribute } from '../../utils/events'
+
+export function parseCreateAccountExecuteResult(
```
Nice, this is going to be pretty helpful already.

dalechyn replied:
Although the createAccount method is available in the AccountFactoryClient itself I will provide the same alias functions that have been declared in the AbstractAccountClient class.

adairrr commented on packages/core/src/actions/public/ans/get-ans-host-address.ts:
```diff
@@ -0,0 +1,41 @@
+  versionControlAddress: string,
+  version?: string,
+) {
+  const registryClient = new VersionControlQueryClient(
+    cosmWasmClient,
+    versionControlAddress,
+  )
+
+  const [ansHostAddress] = await registryClient
+    .modules({
+      infos: [
+        {
+          name: ANS_HOST_MODULE_NAME,
+          namespace: 'abstract',
+          version: version ? { version } : 'latest',
+        } satisfies VersionControlTypes.ModuleInfo,
+      ],
+    })
+    .then(({ modules }) =>
+      modules.map(({ module }) => versionControlModuleToAddress(module)),
+    )
+
+  if (!ansHostAddress) {
+    throw new Error(
+      `Could not get ansHostAddress for version ${version} from registry ${versionControlAddress}`,
+    )
+  }
```
We can create a `getModuleAddressFromVersionControl` method and refrain from repeating the logic for ANS, AF, and MF.

dalechyn replied:
Not sure what the MF is, Module Factory?

The rest is fixed.

adairrr replied:
Yes, though I we actually don't expose the MF functionality.

adairrr commented on packages/core/src/actions/get-accounts-of-owner-from-api.ts:
```diff
@@ -0,0 +1,43 @@
+import { gql } from '../codegen/gql'
+import { accountIdApiFormatToAccountId } from '../utils/account-id/account-id-api-format-to-account-id'
+
+export async function getAccountsOfOwnerFromApi(
```
```suggestion
export async function getAccountsByOwnerFromApi(
```

adairrr commented on packages/core/src/actions/public/manager/get-manager-query-client.ts:
Probably WIP?

adairrr commented on packages/core/src/clients/AbstractClient.ts:
We should update lines 141-165 by removing everything requested except the account id as we're not using any of the other info. (Similar to other comments on over-requesting).

adairrr commented on packages/core/src/clients/AbstractClient.ts:
```diff
@@ -263,16 +258,6 @@ export class AbstractQueryClient implements IAbstractQueryClient {
-   * Query the Abstract API.
-   * @param params
-   */
-  public async queryApi<TResult, V extends Variables = Variables>(
```
Why remove this method? Found it handy as a dev

dalechyn replied:
that file is planned to be deleted, all the legacy support will be in another directory.

adairrr commented on packages/core/src/clients/create-account-public-client.ts:
```diff
@@ -0,0 +1,38 @@
+} from './decorators/account-public.js'
+
+export type AccountPublicClientConfig = ClientConfig & {
+  accountId: VersionControlTypes.AccountId
```
`chainName` is "inferrable" from the `CosmWasmClient` chain-id, but IMO would be better to provide the full `AccountId` type with the chain name for sanity checking that we're getting the account on the correct chain.

dalechyn replied:
I found that rather not useful when implemented the `examples/` as I had to hard-code the chain names myself although it is already "inferrable".

I think it's pretty trivial that by connecting to a `neutron` chain in example, you're pulling `neutron` specific data.

adairrr replied:
Yes, but there are cases (like in the account creation page on the Console) where multiple CosmWasm clients are in use, which could have the dev pass in an incorrect one. Best practice would be, however, to check the chain id of the cosmwasm client, of course.

Something that's missing in the AccountId logic as it stands is that the FE enforced that "local" account ids were never allowed because they're ambiguous, and now we're allowing them again.

adairrr commented on packages/core/src/clients/decorators/public.ts:
```diff
@@ -0,0 +1,62 @@
+  getAccountFactoryAddress(
+    ...args: SliceFirst<Parameters<typeof getAccountFactoryAddress>>
+  ): ReturnType<typeof getAccountFactoryAddress>
+  getAccountFactoryQueryClient(
+    ...args: SliceFirst<Parameters<typeof getAccountFactoryQueryClient>>
+  ): ReturnType<typeof getAccountFactoryQueryClient>
+  getAccountFactoryQueryClientFromApi(
+    ...args: SliceFirstTwo<
+      Parameters<typeof getAccountFactoryQueryClientFromApi>
+    >
+  ): ReturnType<typeof getAccountFactoryQueryClientFromApi>
```
To confirm - we're exposing both the "raw" and "api-based" client retrievals so that we can use this the former in building the Subgraph or just to allow more granular access to devs?

dalechyn replied:
More granular access

adairrr commented on packages/core/src/actions/public/account-factory/get-account-factory-address.ts:
```diff
@@ -0,0 +1,42 @@
+const ACCOUNT_FACTORY_MODULE_NAME = 'account-factory'
+
+export async function getAccountFactoryAddress(
+  cosmWasmClient: CosmWasmClient,
+  versionControlAddress: string,
+  version?: string,
+) {
+  const registryClient = new VersionControlQueryClient(
+    cosmWasmClient,
+    versionControlAddress,
+  )
```
To confirm - We refrain from passing the `VersionControlQueryClient` explicitly instead of the CWClient and the VC Address because we don't want to tie the impl to the ts-codegen version?

dalechyn replied:
The `cosmWasmClient` is propagated by decorators therefore it simplifies the overall API, yet if we would pass such an entity explicitly the developer would have to create one himself.

adairrr commented on packages/core/src/clients/decorators/wallet.ts:
```diff
@@ -0,0 +1,44 @@
+  getAnsHostClientFromApi(
+    ...args: SliceFirstTwo<Parameters<typeof getAnsHostClientFromApi>>
+  ): ReturnType<typeof getAnsHostClientFromApi>
+  getVersionControlClientFromApi(
```
We could expose a `getVersionControlClient` method that still takes in the CWClient and VC Address and just returns the client explicitly. Would make the API consistent.

adairrr commented on packages/core/src/utils/ans/utils/ans-name-to-ibc-token-name.ts:
```diff
@@ -0,0 +1,6 @@
+import { CHAIN_SEPARATOR } from '../constants'
+export function getIbcTokenName(chainName: string, asset: string) {
```
```suggestion
export function getIbcTokenAnsName(chainName: string, asset: string) {
```

adairrr replied:
This is also a question - do we want to clarify that "name" is ANS specific or is "name" of an asset specific enough?

dalechyn replied:
better to clarify

dalechyn replied:
Was renamed to `ibcTokenToAnsId`

adairrr commented on packages/core/src/utils/ans/utils/format-ans-name-without-protocol.ts:
```diff
@@ -0,0 +1,13 @@
+
+export function formatAnsNameWithoutProtocol<
+  const TName extends `${string}${typeof PROTOCOL_SEPARATOR}${string}`,
+>(name: TName): Lowercase<s.Split<TName, typeof PROTOCOL_SEPARATOR>[1]> {
```
These typings are fun.

adairrr commented on packages/core/src/utils/assets/cw20/encode-cw20-asset-send-msg.ts:
```diff
@@ -0,0 +1,24 @@
+  const TAddress extends string = string,
+>(asset: CW20Asset<TAddress>, sender: string, msg: TMsg) {
+  return {
+    typeUrl: '/cosmwasm.wasm.v1.MsgExecuteContract',
```
We should create a task for functionizing the execute contract (and maybe other) message creation as well.

adairrr commented on packages/core/src/utils/tokens/get-token-from-ans.ts:
```diff
@@ -0,0 +1,54 @@
+) {
+  const result = await request(
+    apiURL,
+    gql(
```
```suggestion
    gql(/* GraphQL */
```

adairrr commented on packages/core/src/utils/tokens/get-token-from-ans.ts:
```diff
@@ -0,0 +1,54 @@
+      `
+  query AssetsQuery($chain: ID!, $filter: IdsFilter!) {
+    ans(chain: $chain) {
+      assets(filter: $filter) {
```
Let's make a task to deprecate the `assets` api to be replaced with `tokens`.

adairrr replied:
We can improve it at the same time to return different types based on cw20 or native, getting rid of the overridden "address"

dalechyn replied:
https://linear.app/abstract-sdk/issue/APP-264/refactor-the-asset-field-to-token-on-subgraph

adairrr commented on packages/core/src/utils/version-control/module-id/module-id.ts:
```diff
@@ -0,0 +1,3 @@
+export const MODULE_DELIMITER = ':'
+
+export type ModuleId = `${string}${typeof MODULE_DELIMITER}${string}`
```
Cool!

adairrr commented on packages/core/src/utils/ans/utils/ans-name-to-lp-token-name.ts:
```diff
@@ -0,0 +1,8 @@
+import { joinAssetNames } from './join-asset-names'
+
+export function ansNameToLpTokenName(
```
```suggestion
export function lpTokenName(
```

adairrr replied:
This would be more appropriately named `lpTokenName` based on `stakingContract`.

dalechyn replied:
Renamed to `lpTokenToAnsId`

adairrr commented on packages/core/src/utils/ans/utils/ans-name-to-ibc-token-name.ts:
```diff
@@ -0,0 +1,6 @@
+import { CHAIN_SEPARATOR } from '../constants'
+export function getIbcTokenName(chainName: string, asset: string) {
```
When are we using `getX` vs just `x`? Seems inconsistent between here and `lpTokenName` based on `stakingContract`.

dalechyn replied:
`getX` is now used strictly for actions.

Action name now starts with `get` or with the actual action, like `create`, `remove` etc.

adairrr reviewed (changes requested):
Overall, the changes look really good. I didn't like them at first, but came to enjoy the way that the code read when diving deeper.

Left quite a few comments / questions, though the most major is that we need to add doc comments for all the public-facing api at least. And secondly, we should ensure we're not requesting too much info from the API when unnecessary.

The refactor of the API / Console to use this is going to be quite a PITA... but good that it'll be done at once.

adairrr commented:
We removed `AnsAsset` which actually was not just a difference with the amount field, but rather the difference that it has a name included. This was useful because sometimes you work with the ans assets (like calling into the dex adapter) but also need the resolved ones, like when sending coins along with that message.

Something we'd considered in the past is allowing for passing `AnsAsset`s types in our calls in addition to `Coin` that would do the resolution and transfer message on behalf of the dev.
```
- To eliminate confusion between `AnsAsset` and `Asset` (which has an
  `amount` field), introduce `Token` type.
```

dalechyn committed:
wip

dalechyn commented on packages/core/src/utils/chain-registry/assets.ts:
Bad replace

dalechyn commented on packages/core/src/utils/chain-registry/chains.ts:
Bad replace

dalechyn committed:
wip: with nested

Checkpoint commit to return to if we decide to come back with nested folders approach.

dalechyn committed:
fix: build

Builds succesfully

dalechyn committed:
feat: legacy support, react and cli updates

dalechyn committed:
chore: changesets

dalechyn committed:
fix: lockfile

dalechyn committed:
fix: resolve review comment

https://github.com/AbstractSDK/abstract.js/pull/30\#discussion_r1452847186

dalechyn committed:
fix: resolve review comment

https://github.com/AbstractSDK/abstract.js/pull/30#discussion_r1452870159

dalechyn committed:
fix: resolve review comment

https://github.com/AbstractSDK/abstract.js/pull/30#discussion_r1452871183

dalechyn committed:
fix: resolve review comment

https://github.com/AbstractSDK/abstract.js/pull/30#discussion_r1452871840

dalechyn committed:
nit: drop comment

dalechyn committed:
fix: resolve review comment

https://github.com/AbstractSDK/abstract.js/pull/30#discussion_r1452884551

dalechyn committed:
fix: resolve review comment

https://github.com/AbstractSDK/abstract.js/pull/30#discussion_r1452885338

dalechyn commented:
> We removed `AnsAsset` which actually was not just a difference with the amount field, but rather the difference that it has a name included. This was useful because sometimes you work with the ans assets (like calling into the dex adapter) but also need the resolved ones, like when sending coins along with that message.
> 
> Something we'd considered in the past is allowing for passing `AnsAsset`s types in our calls in addition to `Coin` that would do the resolution and transfer message on behalf of the dev.
> 
> ```
> - To eliminate confusion between `AnsAsset` and `Asset` (which has an
>   `amount` field), introduce `Token` type.
> ```

I addressed that issue, and now `AnsAsset` extends `Asset` by introducing the `name` property.

dalechyn committed:
fix: build

dalechyn committed:
nit: drop unused code

adairrr commented on packages/core/src/actions/account/public/get-base-asset.ts:
```diff
@@ -0,0 +1,17 @@
+    apiUrl,
+  )
+  const { base_asset } = await proxyQueryClient.baseAsset()
+  return base_asset
```
This return will be
```
  | {
      native: string
    }
  | {
      cw20: Addr
    }
``` 
so we should type the return as `Token` if I'm not mistaken

adairrr commented on packages/core/src/actions/account/public/get-module-address.ts:
```diff
@@ -0,0 +1,22 @@
+  accountId: VersionControlTypes.AccountId,
+  cosmWasmClient: CosmWasmClient,
+  apiUrl: string,
+  ...params: Parameters<typeof ManagerQueryClient.prototype.moduleAddresses>
```
These params would be `{ ids }: { ids: string[] }` rather than `id: string` which I'm assuming what we're looking for

dalechyn replied:
good catch!

adairrr commented on packages/core/src/actions/account/public/get-owner.ts:
```diff
@@ -0,0 +1,22 @@
+  accountId: VersionControlTypes.AccountId,
+  cosmWasmClient: CosmWasmClient,
+  apiUrl: string,
+) {
```
```suggestion
): Promise<string | null> {
```

dalechyn replied:
The exact return type is already derived from the function.

adairrr commented on packages/core/src/actions/account/public/get-sub-account-sequences.ts:
```diff
@@ -0,0 +1,22 @@
+  )
+  const { sub_accounts } = await managerQueryClient.subAccountIds(...params)
+
+  return sub_accounts
```
Would be good to transform these into Abstract Account Ids

dalechyn replied:
There is `get-sub-account-ids` that does that.

The original class however had a separate method for resolving sequences.

adairrr reviewed (commented):
Overall comment: we should ensure that we have type the returns of actions to be `ExecuteResult` as well as the query results - which should be typed to what the user would expect.

adairrr commented on packages/core/src/actions/account/wallet/withdraw.ts:
```diff
@@ -0,0 +1,39 @@
+  apiUrl: string,
+  sender: string,
+  assets: Asset[],
+  recipient: string,
```
Do you feel that the positional arguments are better than an object?

adairrr commented on packages/core/src/actions/get-ans-host-address-from-api.ts:
```diff
@@ -0,0 +1,27 @@
+        }
+        chainInfo(chain: $chain) {
+          rpcUrl
+        }
```
```suggestion
        deployment(chain: $chain) {
          ansHost
        }
```

adairrr commented on packages/core/src/actions/public/get-module-address-from-version-control.ts:
```diff
@@ -0,0 +1,46 @@
+
+export enum CommonModuleNames {
+  ACCOUNT_FACTORY_MODULE_NAME = 'account-factory',
+  ANS_HOST_MODULE_NAME = 'ans-host',
```
`_MODULE_NAME` is Extraneous

adairrr commented on packages/core/src/actions/public/get-module-address-from-version-control.ts:
```diff
@@ -0,0 +1,46 @@
+  ANS_HOST_MODULE_NAME = 'ans-host',
+}
+
+export async function getModuleAddressFromVersionControl(
```
This retrieves *abstract* module addresses from version control, as it includes the `abstarct:` namespace in its queries. 

Perhaps `getAbstractModuleAddressFromVersionControl`? Otherwise, we could change the moduleId to be a ModuleId.

adairrr commented on packages/core/src/actions/wallet/create-account.ts:
```diff
@@ -0,0 +1,49 @@
+  >,
+  ...params: SliceFirst<
+    Parameters<typeof AccountFactoryClient.prototype.createAccount>
+  >
```
We should maintain consistency whether we're using object or positional parameter syntax. I prefer the object syntax, as is done here.

adairrr commented on packages/core/src/actions/wallet/get-version-control-client-from-api.ts:
```diff
@@ -0,0 +1,19 @@
+export async function getVersionControlClientFromApi(
+  cosmWasmClient: SigningCosmWasmClient,
+  apiUrl: string,
+  sender: string,
```
Another consistency issue is that the sender moves from 2 to 3 position depending on whether the apiUrl is provided. One would expect it to be the same and the `apiUrl` be 3.

adairrr commented on packages/core/src/clients/decorators/api.ts:
```diff
@@ -0,0 +1,43 @@
+
+export function apiActions(apiUrl: string): ApiActions {
+  return {
+    getAccountFactoryAddressFromApi: (...args) =>
+      getAccountFactoryAddressFromApi(apiUrl, ...args),
+    getAnsHostAddressFromApi: (...args) =>
+      getAnsHostAddressFromApi(apiUrl, ...args),
+    getAccountsByOwnerFromApi: (...args) =>
+      getAccountsByOwnerFromApi(apiUrl, ...args),
+    getAnsTokenFromApi: (...args) => getAnsTokenFromApi(apiUrl, ...args),
+    getAnsTokensFromApi: (...args) => getAnsTokensFromApi(apiUrl, ...args),
+    getVersionControlAddressFromApi: (...args) =>
+      getVersionControlAddressFromApi(apiUrl, ...args),
```
If this is the api client wouldn't it make sense to name them without the `FromApi` postfix?

dalechyn replied:
We could rename them here but overall it's better to keep the postfix

adairrr commented on packages/core/src/utils/account-id/account-id-to-parameter.ts:
```diff
@@ -0,0 +1,5 @@
+import { VersionControlTypes } from 'src/codegen/abstract'
+
+export function accountIdToParameter(id: VersionControlTypes.AccountId) {
+  return { seq: id.seq, trace: id.trace }
+}
```
The parameter should be the `AccountId` type rather than `VersionControlTypes.AccountId`,

adairrr replied:
Unless it's meant just to truncate the `chainName`?

dalechyn replied:
this method does not need `chainName` as it doesn't use one in the response

dalechyn committed:
fix: resolve review comments

dalechyn commented:
> Overall comment: we should ensure that we have type the returns of actions to be `ExecuteResult` as well as the query results - which should be typed to what the user would expect.

All the mutating actions return `ExecuteResult` except for:
- `deposit`: `DeliverTxResponse`
- `execute`: `DeliverTxResponse`
- `withdraw`: `DeliverTxResponse`
- `createAccount`: returns the resolved addresses for a created account

dalechyn committed:
ci: add manual release workflow

dalechyn committed:
ci: workflow fix

dalechyn committed:
ci: workflow fix

adairrr commented:
> > Overall comment: we should ensure that we have type the returns of actions to be `ExecuteResult` as well as the query results - which should be typed to what the user would expect.
> 
> All the mutating actions return `ExecuteResult` except for:
> 
> * `deposit`: `DeliverTxResponse`
> * `execute`: `DeliverTxResponse`
> * `withdraw`: `DeliverTxResponse`
> * `createAccount`: returns the resolved addresses for a created account

Ah yes, forgot about `DeliverTxResponse`... this is because of the use of `signAndBroadcast` if I'm not mistaken. What's interesting is that the `ExecuteResult` doesn't extend `DeliverTxResponse` because the following logic just "removes" fields from it.

```ts
    async executeMultiple(senderAddress, instructions, fee, memo = "") {
...
        const result = await this.signAndBroadcast(senderAddress, msgs, fee, memo);
        if ((0, stargate_1.isDeliverTxFailure)(result)) {
            throw new Error(createDeliverTxResponseErrorMessage(result));
        }
        return {
            logs: stargate_1.logs.parseRawLog(result.rawLog),
            height: result.height,
            transactionHash: result.transactionHash,
            events: result.events,
            gasWanted: result.gasWanted,
            gasUsed: result.gasUsed,
        };
        ```

dalechyn committed:
refactor: change input api for core

adairrr commented on packages/core/src/actions/account/public/get-total-value.ts:
```diff
@@ -1,17 +1,24 @@
+    },
+  })
   const res = await proxyQueryClient.totalValue()
   return +res.total_value
```
We should add a todo here because this value could overflow `number`

adairrr commented on packages/core/src/actions/account/wallet/remove-namespace.ts:
```diff
@@ -1,31 +1,38 @@
+  sender: string
+}>
+
+export async function removeNamespace({
```
Would we be able to rename this `revokeNamespace`

adairrr commented on packages/core/src/actions/get-version-control-address-from-api.ts:
```diff
@@ -1,10 +1,15 @@
+  apiUrl: string
+  chainId: string
+}>
+
```
The graphql query below still requests extraneous info

adairrr commented on packages/core/src/actions/public/get-abstract-module-address-from-version-control.ts:
```diff
@@ -0,0 +1,51 @@
+    },
+  })
+
+  const [accountFactoryAddress] = await versionControlQueryClient
```
nit: incorrect variable name

adairrr commented on packages/core/src/types/with-args.ts:
```diff
@@ -0,0 +1,30 @@
+  args: Prettify<Readonly<TArgs>>
+}
+
+export type WithArgsAndCosmWasmSignOptions<
+  TArgs extends {} | readonly unknown[],
+> = Prettify<
+  WithArgs<TArgs> & {
+    readonly fee: number | StdFee | 'auto'
+    readonly memo?: string
+    readonly funds?: (Coin & { [key: string]: string })[]
+  }
+>
```
🔥

dalechyn committed:
fix: src imports

dalechyn committed:
fix: resolve review comments

dalechyn committed:
Merge remote-tracking branch 'origin/mainline' into refactor/core

adairrr commented on packages/react/src/hooks/use-accounts.ts:
```diff
@@ -47,7 +47,9 @@ export function useAccounts(
 
-    return client.getAccountsByOwnerFromApi(owner, [chain])
+    return client.getAccountsByOwnerFromApi({
+      args: { owner, chains: [chain] },
```
Do we want to allow the user to request accounts from multiple chains? Or is there going to be another hook?

dalechyn replied:
hm let's make it array-value ambitious, second

dalechyn committed:
feat: ambitious array-value chain parameter
